File: solutions/53-maximum-subarray/solution.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Solution(object):
    """
    Intro
    """
    def maxSubArray(self, nums):
        if not nums:
            return None
        maxSum = nums[0]
        curMax = nums[0]
        for i in range(1, len(nums)):
            curMax = max(curMax + nums[i], nums[i])
            maxSum = max(maxSum, curMax)
        return maxSum

File: solutions/53-maximum-subarray/solution.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution
{
public:
    Solution (){};
    virtual ~Solution (){};

    int maxSubArray(vector<int>& nums) {
        int maxSum = 0;
        int curSum = nums[0];
        int i, j;
        for (i = 0; i < nums.size(); ++i) {
            curSum = nums[i];
            maxSum = max(curSum, maxSum);
            for (j = i+1; j < nums.size(); ++j) {
                curSum += nums[j];
                maxSum = max(curSum, maxSum);
            }
        }
        return maxSum;
    }
};

int main(int argc, char *argv[])
{
    vector<int> v = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    int ms = Solution().maxSubArray(v);
    cout << ms << endl;
    return 0;
}
